books/kestrel/number-theory/tonelli-shanks-proof.lisp

```diff
@@ -0,0 +1,1220 @@
+@(thm t-s-aux-base-case)
+<p> Assuming some properties about the loop invariants and using the theorems t-s-aux-loop-invariants, t-s-aux-base-case, we can prove the following theorem about t-s-aux function of the Tonelli-Shanks algorithm. </p>
+@(thm t-s-aux-equiv)
+<p> Proof of soundness of the algorithm: </p>
+@(thm tonelli-shanks-sqrt-aux-is-sqrt-modp)
+<p> Key lemma required to prove that the algorithm is correct: </p>
+@(thm modx^2-y^2)
+<p> Proof that the Tonelli-Shanks modular square root algorithm is correct: </p>
+@(thm tonelli-shanks-sqrt-aux-is-correct)")
+
+(local
+ (encapsulate
+   ()
+   (local (include-book "kestrel/arithmetic-light/mod-and-expt" :dir :system))
+
+   (defthm mod-of-expt-of-mod
+     (implies (and (natp i)
+                   (integerp x)
+                   (posp y))
+              (equal (mod (expt (mod x y) i) y)
+                     (mod (expt x i) y))))
+
+   (defthm mod-of-*-of-expt-of-mod
+     (implies (and (natp i)
+                   (integerp x1)
+                   (integerp x2)
+                   (posp y))
+              (equal (mod (* x1 (expt (mod x2 y) i)) y)
+                     (mod (* x1 (expt x2 i)) y))))
+
+   (defthm natp-2^x
+     (implies (natp x)
+              (natp (expt 2 x)))
+     )
+   ))
+
+(local
+ (encapsulate
+   ()
+   
+   (local (include-book "kestrel/arithmetic-light/expt" :dir :system))
+   (local (include-book "arithmetic/equalities" :dir :system))
+
+   (defthm exponents-add-for-nonneg-exponents
+     (implies (and (<= 0 i)
+                   (<= 0 j)
+                   (integerp i)
+                   (integerp j))
+              (equal (* (expt r i) (expt r j))
+                     (expt r (+ i j)))))
+
+   (defthm expt-half-linear
+     (implies (natp i)
+              (equal (+ (expt 2 (+ -1 i))
+                        (expt 2 (+ -1 i)))
+                     (expt 2 i))))
+
+   (defthm exponents-multiply
+     (implies (and (integerp i)
+                   (integerp j))
+              (equal (expt (expt r i) j)
+                     (expt r (* i j)))))
+
+   (defthm posp!=0
+     (implies (posp x)
+              (not (= x 0))))
+   ))
+
+
+(local
+ (encapsulate
+   ()
+
+   (local
+    (defthm lrs-aux-lemma-1
+      (implies (and (natp tt^2^i)
+                    (posp i)
+                    (natp m)
+                    (natp p)
+                    (< 2 p)
+                    (< i m)
+                    (< (+ i 1) m)
+                    (equal (least-repeated-square-aux i tt^2^i m p) 0))
+               (equal (least-repeated-square-aux (+ i 1) tt^2^i m p) 0))))
+
+   (local
+    (defthm lrs-aux-lemma-2
+      (implies (and (natp tt^2^i)
+                    (posp i)
+                    (natp m)
+                    (natp p)
+                    (< 2 p)
+                    (< i m)
+                    (equal (least-repeated-square-aux i tt^2^i m p) 0))
+               (equal (least-repeated-square-aux (- m 1) tt^2^i m p) 0))
+      :hints (("goal"
```

Comment:
Does lowercase "goal" work?  There are a few instances of this.

Reply:
Yes, it works. I was to do make regression and certify all the files, and I have confirmed that it works with Stephen and Alessandro in one of Friday's meetings.

Reply:
It seems to work with "goal" or "Goal", but not if I take out the hint.  So I guess lowercase "goal" does work!

Reply:
That's good to know. Thanks.

Reply:
Yes, case-insensitivity of the goal-spec is documented behavior.  See
this in :doc hints 

    where [goal-spec] is as described elsewhere (see [goal-spec]) and the

and then see this in :doc goal-spec:

    The matching algorithm is case-insensitive.